cabal-install/src/Distribution/Client/SetupWrapper.hs

```diff
@@ -400,38 +476,60 @@ getSetup verbosity options mpkg = do
 -- | Decide if we're going to be able to do a direct internal call to the
 -- entry point in the Cabal library or if we're going to have to compile
 -- and execute an external Setup.hs script.
-getSetupMethod
+withSetupMethod
   :: Verbosity
   -> SetupScriptOptions
   -> PackageDescription
   -> BuildType
-  -> IO (Version, SetupMethod, SetupScriptOptions)
-getSetupMethod verbosity options pkg buildType'
+  -> AllowInLibrary
+  -> ( forall kind. (Version, SetupMethod kind, SetupScriptOptions ) -> r )
+  -> IO r
+withSetupMethod verbosity options pkg buildType' allowInLibrary with
   | buildType' == Custom
-      || buildType' == Hooks
+      || (buildType' == Hooks && isJust (useLoggingHandle options))
       || maybe False (cabalVersion /=) (useCabalSpecVersion options)
-      || not (cabalVersion `withinRange` useCabalVersion options) =
-      getExternalSetupMethod verbosity options pkg buildType'
-  | isJust (useLoggingHandle options)
-      -- Forcing is done to use an external process e.g. due to parallel
-      -- build concerns.
-      || forceExternalSetupMethod options =
-      return (cabalVersion, SelfExecMethod, options)
-  | otherwise = return (cabalVersion, InternalMethod, options)
-
-runSetupMethod :: WithCallStack (SetupMethod -> SetupRunner)
+      || not (cabalVersion `withinRange` useCabalVersion options)
+      || allowInLibrary == Don'tAllowInLibrary =
+      with <$> getExternalSetupMethod verbosity options pkg buildType'
+  | -- TODO: once we refactor the Cabal library to be able to take a logging
+    -- handle as an argument (e.g. by putting it in Verbosity), we will be able
+    -- to get rid of the self-exec method.
+    isJust (useLoggingHandle options) =
+      return $ with (cabalVersion, SelfExecMethod, options)
+  | otherwise
+  = do
+    abiOK <-
+      if buildType' == Hooks
+      then do
+        -- SetupHooks TODO: getExternalSetupMethod compiles the hooks executable.
+        -- That functionality should be moved here.
```

Comment:
Is there anything which makes this trickier than it sounds?

Reply:
I've since refactored to address this point.